File: leetcode/js/46. Permutations.js
var permute = function (nums) {
  let res = [];
  let visited = Array(nums.length).fill(false);
  dfs(nums, res, [], visited);
  return res;
};

var dfs = function (nums, res = [], curr = [], visited = []) {
  if (curr.length == nums.length) {
    res.push(curr.slice());
    return;
  }
  for (let i in nums) {
    if (visited[i] === false) {
      visited[i] = true;
      curr.push(nums[i]);
      dfs(nums, res, curr, visited);
      curr.pop();
      visited[i] = false;
    }
  }
};

console.log(permute([1, 2, 3]));

File: leetcode/python/46. Permutations.py
def permute(nums):
  res = []
  visited = [False] * len(nums)
  dfs(nums,res,[],visited)
  return res 

def dfs(nums,res=[],curr=[],visited=[]):
  if len(curr) == len(nums):
    res.append(curr[:])
    return
  
  for i in range(len(nums)):
    if visited[i] == False:
      visited[i] = True
      curr.append(nums[i])
      dfs(nums,res,curr,visited)
      curr.pop()
      visited[i] = False

print(permute([1,2,3]))